File: digital_signal_processing/farrow_filter/python/lagrange_genCoeff.py
"""
 lagrange_genCoeff.py
 Author: Daniele Giardino
 Date: 2025.06.24

 Description: 
   Generates the coefficients of the Farrow filter using the Lagrange
   interpolation method.
"""
import numpy as np
from scipy.special import binom

def nchoosek(n, k):
  if k == 0:
    r = 1
  else:
    r = n/k * nchoosek(n-1, k-1)
  return round(r)

def lagrange_genCoeff(num_coeffs):
  """Generates Lagrange coefficients for the modified Farrow structure.

  Args:
    num_coeffs: The number of coefficients in the Farrow structure.

  Returns:
    Q: The matrix of Lagrange coefficients.
  """

  N = num_coeffs - 1  # Filter's order
  M = N + 1  # Number of FIR filters

  # Vandermonde matrix
  U = np.zeros((N + 1, N + 1))
  for i in range(0, N+1):
    for j in range(0, N+1):
      U[j, i] = j**i

  # Compute the inverse of the Vandermonde matrix
  Q = np.linalg.inv(U)

  # Modified Farrow structure coefficients
  T = np.zeros((N + 1, N + 1))
  for n in range(0, N + 1):
    for m in range(0, N + 1):
      if n >= m:
        #coeff_bin = np.math.factorial(n) // (np.math.factorial(m) * np.math.factorial(n - m))
        #coeff_bin = nchoosek(n, m)
        coeff_bin = binom(n, m)
        # print("(%i,%i)=%i" % (n,m,coeff_bin))
        T[n, m] = (np.floor(N / 2)**(n - m)) * coeff_bin

  # Transpose the T matrix
  # Matlab - T = T.'
  T = T.transpose()

  # Calculate the Lagrange coefficients
  # Matlab - Q = T / U
  Q = np.dot(T, np.linalg.inv(U))

  # Delete small values
  val_tol = 1e-15
  for n in range(0, N + 1):
    for m in range(0, N + 1):
      Q[n,m] = 0 if np.abs(Q[n,m])<val_tol else Q[n,m]

  return Q

File: digital_signal_processing/farrow_filter/matlab/lagrange_genCoeff.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% lagrange_genCoeff.m
% Author: Daniele Giardino
% Date: 2025.06.24
%
% Description: 
%   Generates the coefficients of the Farrow filter using the Lagrange
%   interpolation method.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function Q = lagrange_genCoeff(numCoeffs)
% Generate Lagrange coefficients for Farrow structure
%
%   Q = lagrange_genCoeff(numCoeffs)
%
%   Input:
%     numCoeffs - Number of coefficients (order + 1) in the Farrow structure
%
%   Output:
%     Q - Matrix of Lagrange coefficients

%% SECTION 1: Initialization
N = numCoeffs - 1;    % Filter order
M = N + 1;            % Number of FIR subfilters

%% SECTION 2: Construct Vandermonde Matrix
% The Vandermonde matrix is used to solve for the Lagrange interpolation coefficients.
U = zeros(M, M);
for i = 0:N
  for j = 0:N
    U(j + 1, i + 1) = j^i;
  end
end

%% SECTION 3: Compute Inverse of Vandermonde Matrix
% The inverse is used to obtain the Lagrange basis coefficients.
Q = inv(U);

%% SECTION 4: Compute Modified Farrow Structure Coefficients
% T matrix contains the binomial and scaling terms for the Farrow structure.
T = ones(size(Q));
for n = 0:N
  for m = 0:N
    if n >= m
      coeff_bin = nchoosek(n, m); % Binomial coefficient
      T(n + 1, m + 1) = floor(N / 2)^(n - m) * coeff_bin;
    else
      T(n + 1, m + 1) = 0;
    end
  end
end

%% SECTION 5: Transpose T Matrix
T = T';

%% SECTION 6: Calculate Final Lagrange Coefficient Matrix
% The final coefficient matrix Q is obtained by multiplying T and inv(U).
Q = T / U;

end
